New port: RESTinio

Lord-Kamina opened the pull request:
Refactored the Portfile and fixed various bugs that were present in my first attempt. Tried the port with the basic example copy-pasted straight from README.md and it's working as expected.

Commented/disabled tests, samples and benchmarks because they have way too many dependencies that aren't needed for the actual port.

#### Description

RESTinio is a header-only C++14 library that gives you an embedded
HTTP/Websocket server. It is based on standalone version of ASIO and targeted
primarily for asynchronous processing of HTTP-requests.

<!-- Use "skip notification" (surrounded with []) to avoid notifying maintainers -->
<!-- (delete all below for minor changes) -->

###### Type(s)
<!-- update (title contains ": U(u)pdate to"), submission (new Portfile) and CVE Identifiers are auto-detected, replace [ ] with [x] to select -->

- [ ] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on
<!-- Generate version information with this command in shell:
    echo "macOS $(sw_vers -productVersion) $(sw_vers -buildVersion)"; echo "Xcode $(xcodebuild -version | awk '{print $NF}' | tr '\n' ' ')"
-->
macOS 10.13.4
Xcode 9.3

###### Verification <!-- (delete not applicable items) -->
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] checked your Portfile with `port lint`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

ctreleaven commented:
I see that RESTinfo depends on http-parser and asio.  Why not have separate portfiles for each rather than defining them as subports in a single file?

mf2k commented:
I completely agree, they should be separated into 3 different portfiles.

Lord-Kamina commented:
Separated the dependencies into their own ports.

mf2k commented on devel/asio/Portfile:
```diff
@@ -0,0 +1,37 @@
+platforms                   darwin
+
+checksums                   rmd160  08cb85fcc1d7c26fa2805d1b6a0f718d8fefb090 \
+                            sha256  657bf83c883d429259d02a5efe010b63cbb571dc344fd13560151fc2a6b5292c
```
Please add the `size` token.

mf2k commented on devel/asio/Portfile:
```diff
@@ -0,0 +1,37 @@
+
+checksums                   rmd160  08cb85fcc1d7c26fa2805d1b6a0f718d8fefb090 \
+                            sha256  657bf83c883d429259d02a5efe010b63cbb571dc344fd13560151fc2a6b5292c
+depends_build               path:bin/autoconf:autoconf path:bin/automake:automake
```
These should be change to port dependencies.

mf2k commented on www/http-parser/Portfile:
```diff
@@ -0,0 +1,32 @@
+checksums                   rmd160  8ebe6c4b7ea28b40c31b1ef1abbfd7a89f57f7fb \
+                            sha256  6d6a66251dd4236c6e15565849777230f5bc78a92e16eb2d40cec1f7c0f5bbe3
+maintainers                 @Lord-Kamina
+homepage                    https://github.com/nodejs/http-parser
```
This line can be removed because it is the default for the GitHub portgroup.

mf2k commented on www/http-parser/Portfile:
```diff
@@ -0,0 +1,32 @@
+                            It does not make any syscalls nor allocations, it does not buffer data, it can be interrupted at anytime. \
+                            Depending on your architecture, it only requires about 40 bytes of data per message stream (in a web server that is per connection).
+checksums                   rmd160  8ebe6c4b7ea28b40c31b1ef1abbfd7a89f57f7fb \
+                            sha256  6d6a66251dd4236c6e15565849777230f5bc78a92e16eb2d40cec1f7c0f5bbe3
```
Please add the `size` token here too.

mf2k commented on www/http-parser/Portfile:
```diff
@@ -0,0 +1,32 @@
+
+patchfiles                  patch-Makefile.diff
+
+use_autoconf                no
```
This line should be removed since it is the default.

mf2k commented on www/restinio/Portfile:
```diff
@@ -0,0 +1,65 @@
+                            port:http-parser \
+                            path:include/fmt/format.h:libfmt \
+                            path:include/fmt/ostream.h:libfmt \
+                            path:include/fmt/time.h:libfmt
```
These 3 path dependencies should be changed into a single port dependency on libfmt.

Lord-Kamina replied:
@mf2k I made the changes (haven't commited yet) but when trying it out with port -vst, trace mode hides the dependency if I declare it as 
```depends_build port:autoconf```
Any idea what could be going wrong?

mf2k replied:
@Lord-Kamina: Which port are you referring to? If it is asio, then it could be a weird issue with use_autoconf. It is not documented and I don't understand how it works. Maybe it automatically adds the build dependency and you are overriding it. Maybe try using "-append" or just check if it gets added automatically in the first place?

Lord-Kamina commented:
@mf2k it was asio, I did get it fixed before committing though. It seems it had to do with what you said. From what I was told on IRC, ```use_autoconf``` automatically adds the dependencies and then ```autoconf.cmd``` removes them. That, along with not using append resulted in a weird interaction. Moving all the autoconf lines to before declaring dependencies fixed it, though.

mf2k commented on devel/asio/Portfile:
```diff
@@ -0,0 +1,40 @@
+depends_build               port:autoconf port:automake
+depends_lib                 port:boost
+
+patchfiles
```
This empty `patchfiles` line seems unnecessary and so it should be removed.

Lord-Kamina replied:
This is probably leftover from when it was all one Portfile.

mf2k commented on www/http-parser/Portfile:
```diff
@@ -0,0 +1,30 @@
+patchfiles                  patch-Makefile.diff
+
+use_configure               no
+configure.cflags            -DHTTP_PARSER_STRICT=0 DESTDIR=${prefix}
```
Is there a reason to set configure cflags if the configure phase is being skipped?

Lord-Kamina replied:
Maybe I should change it to build.cflags I guess?

mf2k replied:
Sure, if they are honored in the build phase.

mf2k commented:
Sorry for the delay again. I'm not sure where we stand on the PR but I have posted a few more specific comments to look at. 

Is this otherwise working and ready to go?

Lord-Kamina commented:
@mf2k Mostly, yeah. I should probably update it because there's almost certainly one or three newer releases upstream; I've been working on a couple other things though. I'll take a look at it tomorrow or Friday.

Lord-Kamina commented:
Ugh. Checksum errors, I think something went wrong rebasing. Will fix it tomorrow morning.

mf2k commented on devel/asio/Portfile:
```diff
@@ -0,0 +1,39 @@
+long_description            Asio is a cross-platform C++ library for network and low-level I/O programming that provides developers with a \
+                            consistent asynchronous model using a modern C++ approach.
+homepage                    http://www.think-async.com
+maintainers                 @Lord-Kamina
```
Please add your email address here. Also can `openmaintainer` be added? Like this:

`maintainers                 {obfuscated-email-address @Lord-Kamina} openmaintainer`

mf2k commented on www/http-parser/Portfile:
```diff
@@ -0,0 +1,30 @@
+checksums                   rmd160  8ebe6c4b7ea28b40c31b1ef1abbfd7a89f57f7fb \
+                            sha256  6d6a66251dd4236c6e15565849777230f5bc78a92e16eb2d40cec1f7c0f5bbe3 \
+                            size    50744
+maintainers                 @Lord-Kamina
```
Please update this per above comment

mf2k commented on www/http-parser/Portfile:
```diff
@@ -0,0 +1,30 @@
+patchfiles                  patch-Makefile.diff
+
+use_configure               no
+build.args                 CFLAGS="-DHTTP_PARSER_STRICT=0"
```
Minor formatting issue. Line up the CFLAGS by adding a space before it.

mf2k commented on www/restinio/Portfile:
```diff
@@ -0,0 +1,64 @@
+depends_build-append        port:boost port:openssl port:pcre port:zlib port:doxygen
+depends_lib-append          port:asio \
+                            port:http-parser \
+                            port:libfmt
```
Please move the `depends` statements to after the `homepage` line.

mf2k commented on www/restinio/Portfile:
```diff
@@ -0,0 +1,64 @@
+categories                  www devel
+platforms                   darwin
+license                     BSD
+maintainers                 @Lord-Kamina
```
Please update this one too.

mojca commented on devel/asio/Portfile:
```diff
@@ -0,0 +1,39 @@
+                            ASIO_DISABLE_STD_STRING_VIEW=1
+build.dir                   ${worksrcpath}/${github.project}
+
+destroot.target             install
```
This one should be default, I don't know why you would need it.

mojca commented on devel/asio/Portfile:
```diff
@@ -0,0 +1,39 @@
+# -*- coding: utf-8; mode: tcl; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- vim:fenc=utf-8:ft=tcl:et:sw=4:ts=4:sts=4
+PortSystem                  1.0
```
Usually an empty line above and below PortSystem, but that's nitpicking.

mojca commented on www/http-parser/Portfile:
```diff
@@ -0,0 +1,30 @@
+# -*- coding: utf-8; mode: tcl; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- vim:fenc=utf-8:ft=tcl:et:sw=4:ts=4:sts=4
+PortSystem                  1.0
```
Same as above (empty line).

mojca commented on www/http-parser/Portfile:
```diff
@@ -0,0 +1,30 @@
+
+build.target                ""
+build.env-append            PREFIX=${prefix}
+build.dir                   ${worksrcpath}
```
This should be default?

mojca commented on www/http-parser/Portfile:
```diff
@@ -0,0 +1,30 @@
+build.dir                   ${worksrcpath}
+
+destroot.env-append         PREFIX=${prefix}
+destroot.target             install
```
This should be default?

mojca commented on www/http-parser/files/patch-Makefile.diff:
```diff
@@ -0,0 +1,21 @@
+--- Makefile.orig	2018-04-30 11:44:19.000000000 -0300
```
Nitpicking: I like to remove the arbitrary timestamp for the sake of next time (to prevent ending up with some bogus timestamp changes when edits are needed in patches).

mojca commented on www/restinio/Portfile:
```diff
@@ -0,0 +1,64 @@
+license                     BSD
+maintainers                 @Lord-Kamina
+description                 Header-only C++14 library that gives you an embedded HTTP/Websocket server.
+depends_build-append        port:boost port:openssl port:pcre port:zlib port:doxygen
```
Usually one dependency per line.

mojca commented on www/restinio/Portfile:
```diff
@@ -0,0 +1,64 @@
+configure.optflags-delete   -Os
+cmake.generator             Unix Makefiles
+compiler.blacklist          *gcc-3.* *gcc-4.* {*gcc-5.[0-3]} \
+                            {clang < 800} macports-clang-3.4 macports-clang-3.5 macports-clang-3.6 macports-clang-3.7
```
I would be tempted at including the cxx11 1.1 PortGroup and only appending some compilers (we don't have cxx14 PG yet).

mojca commented on www/restinio/files/patch-Doxyfile.diff:
```diff
@@ -0,0 +1,4989 @@
+--- dev/Doxyfile.orig	2018-04-30 18:35:55.000000000 -0300
++++ dev/Doxyfile	2018-04-30 18:38:21.000000000 -0300
+@@ -1,2493 +1,2493 @@
```
Scratching my head a bit. Is this a 5k line patch file which merely converts CRLF to LF? I didn't check, I'm honestly asking. If so, I would kind of prefer to add a dependency on `dos2unix` and run the tool in post-patch phase. Or run some regexp or something (I'm sure there's a more straightforward way that one that requires an additional dependency). In any case storing a huge diff for the sake of lineendings alone seems like an overkill.

Lord-Kamina commented on devel/asio/Portfile:
```diff
@@ -0,0 +1,39 @@
+PortGroup                   github 1.0
+
+github.setup                chriskohlhoff asio 1.12.0 asio-
+git.branch                  [strsed [join ${github.tag_prefix}]${github.version}[join ${github.tag_suffix}] {g/\./-/}]
```
Might be an error on my part somewhere else on the portfile but it's needed because the distfile had a sort-of unexpected name for some reason.

mojca replied:
I figured it out. This was only needed for the fetch phase. The real tag is `asio-1-12-0` rather than `asio-1.12.0` as the `github.setup` suggests. I tried to remove the line after I already had it on my local drive which worked fine, but it failed on Travis. I then changed it to `1-12-0` and used a slightly simpler expression to fix the version number afterwards.

Lord-Kamina commented on devel/asio/Portfile:
```diff
@@ -0,0 +1,39 @@
+
+configure.cmd               ./configure
+configure.cxxflags-append   -std=gnu++14
+configure.args              --prefix=${prefix}
```
Could also be a left-over from when I had everything in a single file but if I added that it was probably because at some point during testing it polluted /usr/local

Lord-Kamina commented:
@mojca I squashed your changes for asio and http-parser since they seemed ok and updated RESTinio while also removing the idiotic patch because apparently it wasn't needed any longer anyway.

mojca commented:
Thank you. A few things still left though:
* That patch is now added in one commit and removed in the other. You probably want to squash those two steps.
* We try to follow the commit message rule to start commits with the port name, so `foo: new port` instead of `new port: foo`. Don't hold my word about this though.

mf2k commented:
`foo: new port` is the accepted format.

Lord-Kamina commented:
Done.

mojca commented:
Thank you. I removed the `"to be squashed"` text from commit messages, added a missing newline at the end of one file and made RESTinio the last commit (it theory you need the dependencies to be available before the main port :). Usually we would have squashed the two asio commits, but since I have a strong personal preference towards github vs. sourceforge it's ok to keep it that way, in particular for the sake of having the code available in archive in case of desire to go back to github one day :)

Thanks for your contribution and sorry to keep you waiting so long.

Lord-Kamina commented:
I also very much prefer github, but sadly personal preference can't compete with missing commits. (And considering that version was released months ago, and an issue has been open upstream at least since May) It seemed unlikely to be fixed any time soon.
